File: ruby/lib/threads/section.rb
# frozen_string_literal: true

require 'digest'
require 'time'

module Threads
  # Section manipulation utilities
  module Section
    class << self
      # Extract section content between ## Name and next ## or EOF
      def extract(content, name)
        pattern = /(?:^|\n)## #{Regexp.escape(name)}\n(.*?)(?=\n## |\z)/m
        match = content.match(pattern)
        return '' unless match

        match[1].strip
      end

      # Replace section content
      def replace(content, name, new_content)
        pattern = /((?:^|\n)## #{Regexp.escape(name)}\n)(.*?)((?=\n## )|\z)/m
        return content unless content.match?(pattern)

        content.gsub(pattern) do
          prefix = ::Regexp.last_match(1)
          suffix = ::Regexp.last_match(3)
          "#{prefix}\n#{new_content}\n\n#{suffix}"
        end
      end

      # Append to section
      def append(content, name, addition)
        section_content = extract(content, name).strip
        new_content = section_content.empty? ? addition : "#{section_content}\n#{addition}"
        replace(content, name, new_content)
      end

      # Ensure section exists, placing before another section
      def ensure_section(content, name, before)
        return content if content.match?(/^## #{Regexp.escape(name)}$/m)

        before_pattern = /(^## #{Regexp.escape(before)})/m
        if content.match?(before_pattern)
          return content.gsub(before_pattern, "## #{name}\n\n\\1")
        end

        # If before section doesn't exist, append at end
        content + "\n## #{name}\n\n"
      end

      # Generate 4-char hash for an item
      def generate_hash(text)
        data = "#{text}#{Time.now.to_f}"
        Digest::MD5.hexdigest(data)[0, 4]
      end

      # Insert log entry with timestamp
      def insert_log_entry(content, entry)
        today = Time.now.strftime('%Y-%m-%d')
        timestamp = Time.now.strftime('%H:%M')
        bullet_entry = "- **#{timestamp}** #{entry}"
        heading = "### #{today}"

        # Check if today's heading exists
        if content.match?(/^### #{Regexp.escape(today)}$/m)
          # Insert after today's heading
          pattern = /(^### #{Regexp.escape(today)}\n)/m
          return content.gsub(pattern, "\\1\n#{bullet_entry}\n")
        end

        # Check if Log section exists
        if content.match?(/^## Log$/m)
          # Insert new heading after ## Log
          return content.gsub(/^## Log$/m, "## Log\n\n#{heading}\n\n#{bullet_entry}")
        end

        # No Log section - append one
        content + "\n## Log\n\n#{heading}\n\n#{bullet_entry}\n"
      end

      # Add note to Notes section with hash comment
      def add_note(content, text)
        content = ensure_section(content, 'Notes', 'Todo')
        hash = generate_hash(text)
        note_entry = "- #{text}  <!-- #{hash} -->"

        # Insert at top of Notes section
        pattern = /(^## Notes\n)/m
        new_content = content.gsub(pattern, "\\1\n#{note_entry}\n")

        [new_content, hash]
      end

      # Add todo item to Todo section
      def add_todo_item(content, text)
        hash = generate_hash(text)
        todo_entry = "- [ ] #{text}  <!-- #{hash} -->"

        # Insert at top of Todo section
        pattern = /(^## Todo\n)/m
        new_content = content.gsub(pattern, "\\1\n#{todo_entry}\n")

        [new_content, hash]
      end

      # Count items matching hash prefix in section
      def count_matching_items(content, section, hash)
        lines = content.split("\n")
        in_section = false
        hash_pattern = "<!-- #{hash}"
        count = 0

        lines.each do |line|
          if line.start_with?("## #{section}")
            in_section = true
          elsif line.start_with?('## ')
            in_section = false
          end

          count += 1 if in_section && line.include?(hash_pattern)
        end

        count
      end

      # Remove line by hash in section
      def remove_by_hash(content, section, hash)
        lines = content.split("\n")
        in_section = false
        hash_pattern = "<!-- #{hash}"
        found = false

        result = lines.reject do |line|
          if line.start_with?("## #{section}")
            in_section = true
          elsif line.start_with?('## ')
            in_section = false
          end

          if in_section && line.include?(hash_pattern) && !found
            found = true
            true
          else
            false
          end
        end

        raise "no item with hash '#{hash}' found" unless found

        result.join("\n")
      end

      # Edit item by hash in section
      def edit_by_hash(content, section, hash, new_text)
        lines = content.split("\n")
        in_section = false
        hash_pattern = "<!-- #{hash}"
        found = false

        result = lines.map do |line|
          if line.start_with?("## #{section}")
            in_section = true
          elsif line.start_with?('## ')
            in_section = false
          end

          if in_section && line.include?(hash_pattern) && !found
            found = true
            # Extract hash from line and rebuild
            match = line.match(/<!--\s*([a-f0-9]{4})\s*-->/)
            if match
              "- #{new_text}  <!-- #{match[1]} -->"
            else
              line
            end
          else
            line
          end
        end

        raise "no item with hash '#{hash}' found" unless found

        result.join("\n")
      end

      # Set todo item checked state by hash
      def set_todo_checked(content, hash, checked)
        lines = content.split("\n")
        in_todo = false
        hash_pattern = "<!-- #{hash}"
        found = false

        result = lines.map do |line|
          if line.start_with?('## Todo')
            in_todo = true
          elsif line.start_with?('## ')
            in_todo = false
          end

          if in_todo && line.include?(hash_pattern) && !found
            found = true
            if checked
              line.sub('- [ ]', '- [x]')
            else
              line.sub('- [x]', '- [ ]')
            end
          else
            line
          end
        end

        raise "no item with hash '#{hash}' found" unless found

        result.join("\n")
      end
    end
  end
end

File: bun/src/section.ts
import * as crypto from 'crypto';
import { formatDate, formatTime } from './utils';

// Extract content of a section (between ## Name and next ## or EOF)
export function extractSection(content: string, name: string): string {
  const escapedName = name.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const pattern = new RegExp(`^## ${escapedName}\\n([\\s\\S]+?)(?=^## |$)`, 'm');
  const match = content.match(pattern);
  if (!match || !match[1]) {
    return '';
  }
  return match[1].trim();
}

// Replace content of a section
export function replaceSection(content: string, name: string, newContent: string): string {
  const escapedName = name.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const pattern = new RegExp(`(^## ${escapedName}\\n)[\\s\\S]+?(^## |$)`, 'gm');

  if (!pattern.test(content)) {
    // Section doesn't exist - handled by caller
    return content;
  }

  // Reset lastIndex since we tested
  pattern.lastIndex = 0;
  // Use function replacement to prevent $ in newContent being interpreted as backreferences
  return content.replace(pattern, (match, p1, p2) => `${p1}\n${newContent}\n\n${p2}`);
}

// Append content to a section
export function appendToSection(content: string, name: string, addition: string): string {
  const sectionContent = extractSection(content, name);
  let newContent = sectionContent.trim();
  if (newContent !== '') {
    newContent += '\n';
  }
  newContent += addition;
  return replaceSection(content, name, newContent);
}

// Ensure a section exists, placing it before another section
export function ensureSection(content: string, name: string, before: string): string {
  const escapedName = name.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const pattern = new RegExp(`^## ${escapedName}`, 'm');
  if (pattern.test(content)) {
    return content;
  }

  const escapedBefore = before.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
  const beforePattern = new RegExp(`(^## ${escapedBefore})`, 'm');

  if (beforePattern.test(content)) {
    return content.replace(beforePattern, `## ${name}\n\n$1`);
  }

  // If before section doesn't exist, append at end
  return content + `\n## ${name}\n\n`;
}

// Generate 4-character hash for an item
export function generateHash(text: string): string {
  const data = `${text}${Date.now()}${Math.random()}`;
  const hash = crypto.createHash('md5').update(data).digest('hex');
  return hash.substring(0, 4);
}

// Insert a timestamped entry to the Log section
export function insertLogEntry(content: string, entry: string): string {
  const now = new Date();
  const today = formatDate(now);
  const timestamp = formatTime(now);
  const bulletEntry = `- **${timestamp}** ${entry}`;
  const heading = `### ${today}`;

  // Check if today's heading exists
  const todayPattern = new RegExp(`^### ${today.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}`, 'm');
  if (todayPattern.test(content)) {
    // Insert after today's heading
    // Use function replacement to prevent $ in entry being interpreted as backreferences
    const pattern = new RegExp(`(^### ${today.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')}\\n)`, 'm');
    return content.replace(pattern, (match, p1) => `${p1}\n${bulletEntry}\n`);
  }

  // Check if Log section exists
  const logPattern = /^## Log/m;
  if (logPattern.test(content)) {
    // Insert new heading after ## Log
    // Use function replacement to prevent $ in entry being interpreted as backreferences
    return content.replace(logPattern, () => `## Log\n\n${heading}\n\n${bulletEntry}`);
  }

  // No Log section - append one
  return content + `\n## Log\n\n${heading}\n\n${bulletEntry}\n`;
}

// Add a note to the Notes section with a hash comment
export function addNote(content: string, text: string): { content: string; hash: string } {
  // Ensure Notes section exists
  content = ensureSection(content, 'Notes', 'Todo');

  const hash = generateHash(text);
  const noteEntry = `- ${text}  <!-- ${hash} -->`;

  // Insert at top of Notes section
  // Use function replacement to prevent $ in text being interpreted as backreferences
  const pattern = /^## Notes\n/m;
  const newContent = content.replace(pattern, () => `## Notes\n\n${noteEntry}\n`);

  return { content: newContent, hash };
}

// Remove a line containing the specified hash comment from a section
export function removeByHash(content: string, section: string, hash: string): string {
  const lines = content.split('\n');
  let inSection = false;
  const hashPattern = `<!-- ${hash}`;
  let found = false;

  const result: string[] = [];
  for (const line of lines) {
    if (line.startsWith(`## ${section}`)) {
      inSection = true;
    } else if (line.startsWith('## ')) {
      inSection = false;
    }

    if (inSection && line.includes(hashPattern) && !found) {
      found = true;
      continue; // skip this line
    }
    result.push(line);
  }

  if (!found) {
    throw new Error(`no item with hash '${hash}' found`);
  }

  return result.join('\n');
}

// Edit item text by hash
export function editByHash(content: string, section: string, hash: string, newText: string): string {
  const lines = content.split('\n');
  let inSection = false;
  const hashPattern = `<!-- ${hash}`;
  let found = false;

  const hashCommentRe = /<!--\s*([a-f0-9]{4})\s*-->/;
  const result: string[] = [];

  for (const line of lines) {
    if (line.startsWith(`## ${section}`)) {
      inSection = true;
    } else if (line.startsWith('## ')) {
      inSection = false;
    }

    if (inSection && line.includes(hashPattern) && !found) {
      found = true;
      // Extract hash from line and rebuild
      const match = line.match(hashCommentRe);
      if (match) {
        result.push(`- ${newText}  <!-- ${match[1]} -->`);
        continue;
      }
    }
    result.push(line);
  }

  if (!found) {
    throw new Error(`no item with hash '${hash}' found`);
  }

  return result.join('\n');
}

// Add a todo item to the Todo section
export function addTodoItem(content: string, text: string): { content: string; hash: string } {
  const hash = generateHash(text);
  const todoEntry = `- [ ] ${text}  <!-- ${hash} -->`;

  // Insert at top of Todo section
  // Use function replacement to prevent $ in text being interpreted as backreferences
  const pattern = /^## Todo\n/m;
  const newContent = content.replace(pattern, () => `## Todo\n\n${todoEntry}\n`);

  return { content: newContent, hash };
}

// Set todo item's checked state by hash
export function setTodoChecked(content: string, hash: string, checked: boolean): string {
  const lines = content.split('\n');
  let inTodo = false;
  const hashPattern = `<!-- ${hash}`;
  let found = false;

  const result: string[] = [];
  for (let line of lines) {
    if (line.startsWith('## Todo')) {
      inTodo = true;
    } else if (line.startsWith('## ')) {
      inTodo = false;
    }

    if (inTodo && line.includes(hashPattern) && !found) {
      found = true;
      if (checked) {
        line = line.replace('- [ ]', '- [x]');
      } else {
        line = line.replace('- [x]', '- [ ]');
      }
    }
    result.push(line);
  }

  if (!found) {
    throw new Error(`no item with hash '${hash}' found`);
  }

  return result.join('\n');
}

// Count items matching a hash prefix in a section
export function countMatchingItems(content: string, section: string, hash: string): number {
  const lines = content.split('\n');
  let inSection = false;
  const hashPattern = `<!-- ${hash}`;
  let count = 0;

  for (const line of lines) {
    if (line.startsWith(`## ${section}`)) {
      inSection = true;
    } else if (line.startsWith('## ')) {
      inSection = false;
    }

    if (inSection && line.includes(hashPattern)) {
      count++;
    }
  }

  return count;
}
